internal/grpctest/tlogger.go

```diff
@@ -67,72 +67,118 @@ func log(t *testing.T, format string, fatal bool, args ...interface{}) {
 		return
 	}
 	if format == "" {
-		if fatal {
+		switch ltype {
+		case errorLog:
+			// fmt.Sprintln is used rather than fmt.Sprint because t.Log uses fmt.Sprintln behavior.
+			if logger.expected(fmt.Sprintln(args...)) {
+				t.Log(args...)
+			} else {
+				t.Error(args...)
+			}
+		case fatalLog:
 			panic(fmt.Sprint(args...))
-		} else {
+		default:
 			t.Log(args...)
 		}
 	} else {
-		if fatal {
-			panic(fmt.Sprintf("%v "+format, args...))
-		} else {
-			t.Logf("%v "+format, args...)
+		format = "%v " + format
+		switch ltype {
+		case errorLog:
+			if logger.expected(fmt.Sprintf(format, args...)) {
+				t.Logf(format, args...)
+			} else {
+				t.Errorf(format, args...)
+			}
+		case fatalLog:
+			panic(fmt.Sprintf(format, args...))
+		default:
+			t.Logf(format, args...)
 		}
 	}
 }
 
 // Update updates the testing.T that the testing logger logs to. Should be done
-// before every test.
+// before every test. It also initializes the tLogger if it has not already.
 func Update(t *testing.T) {
+	if logger == nil {
+		logger = &tLogger{v: 0}
+		vLevel := os.Getenv("GRPC_GO_LOG_VERBOSITY_LEVEL")
+		if vl, err := strconv.Atoi(vLevel); err == nil {
+			logger.v = vl
+		}
+		grpclog.SetLoggerV2(logger)
+	}
 	logger.t = t
+	logger.errors = nil
+}
+
+// Expect declares an error to be expected. For the next test, all error logs
+// matching the expression (using FindString) will not cause the test to fail.
+// "For the next test" is includes all the time until the next call to Update().
+func Expect(expr string) {
+	re, err := regexp.Compile(expr)
+	if err != nil {
+		logger.t.Error(err)
+		return
+	}
+	logger.errors = append(logger.errors, re)
+}
+
+func (g *tLogger) expected(s string) bool {
+	for _, re := range g.errors {
+		if re.FindStringIndex(s) != nil {
+			return true
```

Comment:
It might be nice to:

1. Remove this error when the match is found.  If needed, add `ExpectErrorN(string, int)` for convenience.
2. In Teardown: `if len(g.errors) != 0 { t.Error; g.errors = nil }`

Hopefully the number of occurrences of error logs is predictable!

Reply:
Added ExpectErrorN. Errors is not a map (mapping to the counts) to accompany this change. Counts decremented as errors are encountered, and if the map is non-empty in teardown it errors.